File: frontend/src/ward.js
// Ward Representative Console Logic
let token = null;
let currentUser = null;
let departments = [];
const issuesContainer = document.getElementById('ward-issues-container');

// Auth Guard Check
async function initializeWard() {
  token = auth.getToken();
  currentUser = await auth.checkAuthGuard(['ward_member']);
  
  if (currentUser) {
    if (currentUser.is_approved === false) {
      document.getElementById('pending-overlay').style.display = 'flex';
      return;
    }
    // Populate profile card
    document.getElementById('profile-name').textContent = currentUser.full_name;
    const initials = currentUser.full_name
      .split(' ')
      .map(n => n[0])
      .join('')
      .toUpperCase();
    document.getElementById('profile-initials').textContent = initials.substring(0, 2);

    // Update subtitle if ward info is available
    if (currentUser.ward_name) {
      document.getElementById('ward-subtitle').textContent = `Review complaints routed to ${currentUser.ward_name}`;
    }
    
    // Load Departments and Ward Queue
    await loadDepartments();
    await loadWardIssues();
    
    // Load Profile & Stats in background
    await loadProfile();
    await loadStats();
  }
}

// Tab Switching
function switchTab(tab) {
  // Hide all tab contents
  document.querySelectorAll('.tab-content').forEach(el => el.style.display = 'none');
  // Deactivate all nav links
  document.querySelectorAll('.sidebar-nav .nav-link').forEach(el => el.classList.remove('active'));
  
  // Show target tab
  document.getElementById(`tab-${tab}`).style.display = 'block';
  // Activate target link
  document.getElementById(`tab-link-${tab}`).classList.add('active');
  
  // Trigger specific reloads
  if (tab === 'analytics') {
    loadStats();
  } else if (tab === 'profile') {
    loadProfile();
  } else if (tab === 'queue') {
    loadWardIssues();
  }
}

// Fetch departments list
async function loadDepartments() {
  try {
    const res = await fetch('/api/departments');
    const data = await res.json();
    if (res.ok && data.success) {
      departments = data.data;
    }
  } catch (err) {
    console.error('Failed to load departments list:', err);
  }
}

// Show Alerts
function showAlert(text, isError = false) {
  const banner = document.getElementById('alert-banner');
  const textEl = document.getElementById('alert-text');
  if (banner && textEl) {
    textEl.textContent = text;
    banner.className = `alert-banner ${isError ? 'alert-error' : 'alert-success'}`;
    banner.style.display = 'flex';
    setTimeout(() => { banner.style.display = 'none'; }, 5000);
  }
}

// Load ward-assigned complaints
async function loadWardIssues() {
  try {
    const res = await fetch('/api/ward/issues', {
      headers: { 'Authorization': `Bearer ${token}` }
    });
    const data = await res.json();
    if (res.ok && data.success) {
      renderWardIssues(data.data);
    } else {
      issuesContainer.innerHTML = `<p style="text-align: center; color: var(--text-muted); padding: 40px;">Failed to load ward queue.</p>`;
    }
  } catch (err) {
    console.error('Error loading ward issues:', err);
    issuesContainer.innerHTML = `<p style="text-align: center; color: var(--text-muted); padding: 40px;">Connection failure.</p>`;
  }
}

// Render Ward Issues
function renderWardIssues(issues) {
  if (!issues || issues.length === 0) {
    issuesContainer.innerHTML = `
      <div style="text-align: center; color: var(--text-muted); padding: 60px;">
        <span style="font-size: 3rem; display: block; margin-bottom: 16px;"></span>
        <p data-i18n="no_ward_issues">No issues routed to your ward.</p>
      </div>
    `;
    if (window.i18n) window.i18n.translatePage();
    return;
  }

  issuesContainer.innerHTML = issues.map(issue => {
    const statusText = window.i18n ? window.i18n.t(`badge-${issue.status}`, issue.status) : issue.status;
    const priorityText = window.i18n ? window.i18n.t(`badge-priority-${issue.priority}`, issue.priority) : issue.priority;
    const formattedDate = new Date(issue.created_at).toLocaleString();

    // Parse uploaded media
    const mediaHTML = (issue.media_urls || []).map(media => {
      if (media.type === 'image') {
        return `<img src="${media.url}" class="gallery-image" onclick="window.open('${media.url}')">`;
      } else if (media.type === 'audio') {
        return `<audio controls style="max-width: 100%; height: 32px; margin-top: 6px;"><source src="${media.url}"></audio>`;
      } else if (media.type === 'video') {
        return `<video controls style="max-width: 100%; max-height: 140px; border-radius: 6px; margin-top: 6px;"><source src="${media.url}"></video>`;
      }
      return '';
    }).join('');

    // Generate redirect options
    const redirectOptions = departments.map(d => 
      `<option value="${d.id}" ${issue.department_id === d.id ? 'selected' : ''}>${d.name}</option>`
    ).join('');

    return `
      <div class="issue-card ${issue.status}">
        <div class="issue-header">
          <div>
            <span class="badge badge-${issue.status}">${statusText}</span>
            <span class="badge badge-priority-${issue.priority}">${priorityText}</span>
          </div>
          <span style="font-size: 0.8rem; color: var(--text-muted);">${formattedDate}</span>
        </div>
        
        <h3 class="issue-title">${escapeHtml(issue.title)}</h3>
        <p class="issue-desc">${escapeHtml(issue.description)}</p>
        
        <div style="font-size: 0.85rem; margin: 12px 0; color: var(--text-secondary); display: flex; flex-direction: column; gap: 4px;">
          <div><strong>Address:</strong> ${escapeHtml(issue.address || 'GPS Coordinates Only')}</div>
          <div><strong>Category:</strong> ${escapeHtml(issue.category)}</div>
          <div><strong>Upvotes:</strong> ${issue.votes ? issue.votes.upvotes : 0}</div>
        </div>

        <!-- Render captured files -->
        ${mediaHTML ? `<div class="media-gallery" style="margin-top: 12px; gap: 8px;">${mediaHTML}</div>` : ''}

        <div class="issue-footer" style="border-top: 1px solid var(--border); padding-top: 12px; margin-top: 12px; display: flex; flex-direction: column; gap: 12px;">
          <!-- Actions panel -->
          <div style="display: flex; gap: 16px; align-items: center; justify-content: space-between; flex-wrap: wrap;">
            
            <div style="display: flex; align-items: center; gap: 8px;">
              <span data-i18n="redirect_dept">Redirect Department:</span>
              <select id="redirect-dept-${issue.id}" style="padding: 6px 12px; border-radius: 6px; border: 1px solid var(--border); background: var(--bg-secondary); color: #fff;">
                ${redirectOptions}
              </select>
              <button onclick="redirectDept('${issue.id}')" class="btn" style="padding: 6px 12px; font-size: 0.8rem;" data-i18n="redirect">Redirect</button>
            </div>
            
          </div>
        </div>
      </div>
    `;
  }).join('');

  if (window.i18n) window.i18n.translatePage();
}

// Set Priority Level
async function setPriority(issueId, level) {
  try {
    const res = await fetch(`/api/ward/issues/${issueId}/priority`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify({ priority: level })
    });
    if (res.ok) {
      showAlert(`Priority updated to ${level}!`, false);
      await loadWardIssues();
    } else {
      showAlert('Failed to update priority level.', true);
    }
  } catch (err) {
    showAlert('Connection error during priority change.', true);
  }
}

// Redirect Department
async function redirectDept(issueId) {
  const select = document.getElementById(`redirect-dept-${issueId}`);
  const deptId = parseInt(select.value);

  try {
    const res = await fetch(`/api/ward/issues/${issueId}/redirect`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify({ department_id: deptId })
    });
    if (res.ok) {
      showAlert('Issue redirected successfully!', false);
      await loadWardIssues();
    } else {
      showAlert('Failed to redirect issue.', true);
    }
  } catch (err) {
    showAlert('Connection error during department redirection.', true);
  }
}

// Load Ward Statistics
async function loadStats() {
  try {
    const res = await fetch('/api/ward/stats', {
      headers: { 'Authorization': `Bearer ${token}` }
    });
    const data = await res.json();
    if (res.ok && data.success) {
      const stats = data.data;
      document.getElementById('stats-total').textContent = stats.total_issues;
      document.getElementById('stats-speed').textContent = stats.avg_res_time_hours > 0 ? stats.avg_res_time_hours : '0.0';

      // Status breakdown render
      const statusList = document.getElementById('stats-status-list');
      const statuses = ['pending', 'in_progress', 'resolved', 'rejected'];
      statusList.innerHTML = statuses.map(s => {
        const count = stats.status_breakdown[s] || 0;
        const pct = stats.total_issues > 0 ? Math.round((count / stats.total_issues) * 100) : 0;
        return `
          <div>
            <div style="display: flex; justify-content: space-between; font-size: 0.8rem; margin-bottom: 4px;">
              <span style="text-transform: capitalize;">${s.replace('_', ' ')}</span>
              <span>${count} (${pct}%)</span>
            </div>
            <div style="background: rgba(255,255,255,0.05); height: 8px; border-radius: 99px; overflow: hidden;">
              <div style="background: var(--primary); width: ${pct}%; height: 100%; border-radius: 99px;"></div>
            </div>
          </div>
        `;
      }).join('');

      // Priority distribution render
      const priorityList = document.getElementById('stats-priority-list');
      const priorities = ['low', 'medium', 'high', 'critical'];
      priorityList.innerHTML = priorities.map(p => {
        const count = stats.priority_breakdown[p] || 0;
        const pct = stats.total_issues > 0 ? Math.round((count / stats.total_issues) * 100) : 0;
        let color = 'var(--text-secondary)';
        if (p === 'high') color = 'var(--warning)';
        else if (p === 'critical') color = 'var(--danger)';
        else if (p === 'low') color = 'var(--success)';
        
        return `
          <div>
            <div style="display: flex; justify-content: space-between; font-size: 0.8rem; margin-bottom: 4px;">
              <span style="text-transform: capitalize;">${p}</span>
              <span>${count} (${pct}%)</span>
            </div>
            <div style="background: rgba(255,255,255,0.05); height: 8px; border-radius: 99px; overflow: hidden;">
              <div style="background: ${color}; width: ${pct}%; height: 100%; border-radius: 99px;"></div>
            </div>
          </div>
        `;
      }).join('');

      // Categories distribution render
      const categoriesList = document.getElementById('stats-categories-list');
      const cats = Object.entries(stats.top_categories);
      if (cats.length === 0) {
        categoriesList.innerHTML = `<p style="font-size: 0.85rem; color: var(--text-secondary); text-align: center; padding: 16px;">No category statistics available.</p>`;
      } else {
        categoriesList.innerHTML = cats.map(([cat, count]) => {
          return `
            <div style="display: flex; justify-content: space-between; font-size: 0.85rem; padding: 8px 12px; background: rgba(255,255,255,0.02); border-radius: 6px; border: 1px solid var(--border);">
              <span style="font-weight: 500;">${escapeHtml(cat)}</span>
              <span style="color: var(--primary); font-weight: 700;">${count}</span>
            </div>
          `;
        }).join('');
      }
    }
  } catch (err) {
    console.error('Error fetching ward stats:', err);
  }
}

// Load Ward Profile
async function loadProfile() {
  try {
    const res = await fetch('/api/ward/profile', {
      headers: { 'Authorization': `Bearer ${token}` }
    });
    const data = await res.json();
    if (res.ok && data.success) {
      const p = data.data;
      document.getElementById('profile-username').value = p.username;
      document.getElementById('profile-fullname').value = p.full_name;
      document.getElementById('profile-contact').value = p.contact || '';
      
      // Update Audit Cards
      document.getElementById('audit-jurisdiction').textContent = p.ward_name;
      
      const ipfsLink = document.getElementById('audit-ipfs-link');
      if (p.ipfs_cid) {
        ipfsLink.textContent = p.ipfs_cid;
        ipfsLink.href = `https://gateway.pinata.cloud/ipfs/${p.ipfs_cid}`;
      } else {
        ipfsLink.textContent = 'None';
        ipfsLink.removeAttribute('href');
      }

      document.getElementById('audit-db-hash').textContent = p.blockchain_hash || 'None';
      document.getElementById('audit-chain-hash').textContent = p.onchain_hash || 'None';

      const badge = document.getElementById('audit-status-badge');
      if (p.is_verified) {
        badge.className = 'status-badge approved';
        badge.textContent = 'Verified (On-Chain)';
      } else if (p.blockchain_hash) {
        badge.className = 'status-badge rejected';
        badge.textContent = 'Verification Failed';
      } else {
        badge.className = 'status-badge pending';
        badge.textContent = 'Unanchored';
      }
    }
  } catch (err) {
    console.error('Error fetching ward profile:', err);
  }
}

// Update Profile Form action
async function updateProfile(event) {
  event.preventDefault();
  const saveBtn = document.getElementById('btn-save-profile');
  saveBtn.disabled = true;
  saveBtn.textContent = 'Updating...';

  const payload = {
    full_name: document.getElementById('profile-fullname').value.trim(),
    contact: document.getElementById('profile-contact').value.trim()
  };

  try {
    const res = await fetch('/api/ward/profile', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify(payload)
    });
    const data = await res.json();
    
    if (res.ok && data.success) {
      showAlert('Profile details successfully re-anchored on blockchain!', false);
      // Update UI elements
      document.getElementById('profile-name').textContent = payload.full_name;
      currentUser.full_name = payload.full_name;
      
      await loadProfile();
    } else {
      showAlert(data.detail || 'Failed to update profile.', true);
    }
  } catch (err) {
    showAlert('Server connection error during profile update.', true);
  } finally {
    saveBtn.disabled = false;
    saveBtn.textContent = 'Save Changes & Update IPFS';
  }
}

// Escape utilities
function escapeHtml(value) {
  if (!value) return '';
  return String(value)
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
    .replaceAll("'", '&#039;');
}

// Startup
window.addEventListener('load', initializeWard);
window.addEventListener('languageChanged', loadWardIssues);
window.auth_tabs = { switchTab }; // export tab switching globally
window.switchTab = switchTab;

File: backend/app/routes/ward.py
import logging
from fastapi import APIRouter, Depends, HTTPException

from backend.app.database import get_connection
from backend.app.auth import RoleChecker, UserResponse
from backend.app.models import ProfileUpdateRequest, PriorityRequest, RedirectRequest
from backend.app.helpers import anchor_ward_profile, _serialize_issue, _assign_dynamic_priorities
import backend.app.blockchain_service as blockchain_service

router = APIRouter(prefix="/api/ward", tags=["ward"])

@router.get("/profile")
async def get_ward_profile(
    current_user: UserResponse = Depends(RoleChecker(["ward_member"]))
) -> dict:
    try:
        with get_connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(
                    """
                    SELECT w.name as ward_name, w.ipfs_cid, w.blockchain_hash, w.id as ward_id,
                           u.full_name, u.contact, u.username
                    FROM users u
                    JOIN wards w ON u.id = w.ward_member_id
                    WHERE u.id = %s
                    """,
                    (current_user.id,)
                )
                row = cursor.fetchone()
                if not row:
                    raise HTTPException(status_code=404, detail="Ward profile not found")
        
        onchain_hash = blockchain_service.get_personnel_hash(current_user.id)
        db_hash = row["blockchain_hash"]
        is_verified = bool(db_hash and onchain_hash and db_hash.lower() == onchain_hash.lower())

        return {
            "success": True,
            "data": {
                "username": row["username"],
                "full_name": row["full_name"],
                "contact": row["contact"],
                "ward_id": row["ward_id"],
                "ward_name": row["ward_name"],
                "ipfs_cid": row["ipfs_cid"],
                "blockchain_hash": row["blockchain_hash"],
                "onchain_hash": onchain_hash,
                "is_verified": is_verified
            }
        }
    except HTTPException:
        raise
    except Exception as exc:
        logging.exception("Failed to fetch ward profile")
        raise HTTPException(status_code=500, detail="Internal server error")

@router.post("/profile")
async def update_ward_profile(
    req: ProfileUpdateRequest,
    current_user: UserResponse = Depends(RoleChecker(["ward_member"]))
) -> dict:
    try:
        with get_connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(
                    "UPDATE users SET full_name = %s, contact = %s WHERE id = %s",
                    (req.full_name, req.contact, current_user.id)
                )
                cursor.execute("SELECT id FROM wards WHERE ward_member_id = %s", (current_user.id,))
                ward_row = cursor.fetchone()
                if not ward_row:
                    raise HTTPException(status_code=404, detail="Ward not found for this user")
                ward_id = ward_row["id"]
            conn.commit()

        await anchor_ward_profile(current_user.id, current_user.username, req.full_name, req.contact or "", ward_id)
        return {"success": True, "message": "Profile updated and anchored to blockchain successfully"}
    except HTTPException:
        raise
    except Exception as exc:
        logging.exception("Failed to update ward profile")
        raise HTTPException(status_code=500, detail="Internal server error")

@router.get("/stats")
async def get_ward_stats(
    current_user: UserResponse = Depends(RoleChecker(["ward_member"]))
) -> dict:
    try:
        with get_connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute("SELECT id FROM wards WHERE ward_member_id = %s", (current_user.id,))
                ward_row = cursor.fetchone()
                if not ward_row:
                    raise HTTPException(status_code=404, detail="Ward not found for this member")
                ward_id = ward_row["id"]

                cursor.execute("SELECT COUNT(*) as count FROM issues WHERE ward_id = %s", (ward_id,))
                total_issues = cursor.fetchone()["count"]

                cursor.execute(
                    "SELECT status, COUNT(*) as count FROM issues WHERE ward_id = %s GROUP BY status",
                    (ward_id,)
                )
                status_rows = cursor.fetchall()
                status_breakdown = {r["status"]: r["count"] for r in status_rows}

                cursor.execute(
                    "SELECT priority, COUNT(*) as count FROM issues WHERE ward_id = %s GROUP BY priority",
                    (ward_id,)
                )
                priority_rows = cursor.fetchall()
                priority_breakdown = {r["priority"]: r["count"] for r in priority_rows}

                cursor.execute(
                    """
                    SELECT AVG(EXTRACT(EPOCH FROM (h.created_at - i.created_at))) / 3600.0 as avg_time
                    FROM issues i
                    JOIN issue_status_history h ON i.id = h.issue_id
                    WHERE i.ward_id = %s AND h.new_status = 'resolved'
                    """,
                    (ward_id,)
                )
                avg_time_row = cursor.fetchone()
                avg_res_time = round(avg_time_row["avg_time"], 1) if avg_time_row and avg_time_row["avg_time"] is not None else 0.0

                cursor.execute(
                    """
                    SELECT category, COUNT(*) as count 
                    FROM issues 
                    WHERE ward_id = %s 
                    GROUP BY category 
                    ORDER BY count DESC 
                    LIMIT 5
                    """,
                    (ward_id,)
                )
                cat_rows = cursor.fetchall()
                top_categories = {r["category"]: r["count"] for r in cat_rows}

        return {
            "success": True,
            "data": {
                "total_issues": total_issues,
                "status_breakdown": status_breakdown,
                "priority_breakdown": priority_breakdown,
                "avg_res_time_hours": avg_res_time,
                "top_categories": top_categories
            }
        }
    except HTTPException:
        raise
    except Exception as exc:
        logging.exception("Failed to fetch ward statistics")
        raise HTTPException(status_code=500, detail="Internal server error")

@router.get("/issues")
async def get_ward_issues(current_user: UserResponse = Depends(RoleChecker(["ward_member"]))) -> dict:
    try:
        with get_connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(
                    """
                    SELECT i.id, i.title, i.description, i.category, i.area, i.address,
                           i.latitude, i.longitude, i.reporter_name, i.contact,
                           i.image_url, i.status, i.created_at, i.hash,
                           i.priority, i.ward_id, w.name as ward_name,
                           i.department_id, d.name as department_name,
                           i.ipfs_cid, i.media_urls, i.completion_proof_ipfs_cid, i.completion_hash,
                           i.upvote_count, i.downvote_count
                    FROM issues i
                    JOIN wards w ON i.ward_id = w.id
                    LEFT JOIN departments d ON i.department_id = d.id
                    WHERE w.ward_member_id = %s
                    ORDER BY i.upvote_count DESC, i.created_at DESC
                    """,
                    (current_user.id,)
                )
                rows = cursor.fetchall()

        items = [_serialize_issue(row) for row in rows]
        items = _assign_dynamic_priorities(items)
        return {"success": True, "count": len(items), "data": items}
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Failed to fetch ward issues: {exc}")

@router.patch("/issues/{issue_id}/priority")
async def update_issue_priority(
    issue_id: str,
    req: PriorityRequest,
    current_user: UserResponse = Depends(RoleChecker(["ward_member"]))
) -> dict:
    raise HTTPException(status_code=403, detail="Priority updates are handled automatically by user votes.")

@router.post("/issues/{issue_id}/redirect")
async def redirect_issue(
    issue_id: str,
    req: RedirectRequest,
    current_user: UserResponse = Depends(RoleChecker(["ward_member"]))
) -> dict:
    try:
        with get_connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(
                    """
                    SELECT i.id FROM issues i
                    JOIN wards w ON i.ward_id = w.id
                    WHERE i.id = %s AND w.ward_member_id = %s
                    """,
                    (issue_id, current_user.id)
                )
                if not cursor.fetchone():
                    raise HTTPException(status_code=403, detail="Not authorized to route issues in another ward")

                cursor.execute("SELECT id FROM departments WHERE id = %s", (req.department_id,))
                if not cursor.fetchone():
                    raise HTTPException(status_code=404, detail="Department not found")

                cursor.execute(
                    "UPDATE issues SET department_id = %s WHERE id = %s",
                    (req.department_id, issue_id)
                )
            conn.commit()
        return {"success": True, "message": "Issue redirected to new department"}
    except HTTPException:
        raise
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Failed to redirect issue: {exc}")
